File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <stdlib.h>

int main(void)
{
    long credit = get_long("Number: ");
    int length = floor(log10(labs(credit))) + 1;
    // printf("%d\n", length);
    if (length == 13 || length == 15 || length == 16) // If the length is valid
    {
        int digits[length];
        for (int i = length - 1; i >= 0; i--)
        {
            digits[i] = credit % 10;
            credit = credit / 10;
        }

        int total = 0;
        for (int i = length - 2; i >= 0; i -= 2) // starting iteration with second-to-last
        {
            int product = digits[i] * 2;
            int length_of_product = floor(log10(abs(product))) + 1;
            if (length_of_product > 1)
            {
                for (int j = length - 1; j >= 0; j--)
                {
                    total += product % 10;
                    product = product / 10;
                }
            }
            else
            {
                total += product;
            }
        }

        for (int i = length - 1; i >= 0; i -= 2)
        {
            total += digits[i];
        }
        // printf("%d\n", total); // right now we have the expected value 13 for test input: 4003600000000014

        if (total % 10 != 0)
        {
            printf("INVALID\n");
            return 0;
        }

        if (length == 15 && digits[0] == 3 && (digits[1] == 4 || digits[1] == 7))
        {
            printf("AMEX\n");
        }
        else if ((length == 13 || length == 16) && digits[0] == 4)
        {
            printf("VISA\n");
        }
        else if (length == 16 && digits[0] == 5 && (digits[1] == 1 || digits[1] == 2 || digits[1] == 3 || digits[1] == 4 || digits[1] == 5))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
        return 0;
    }
    else // if length is invalid
    {
        printf("INVALID\n");
        return 0;
    }
}

// American Express uses 15-digit numbers | 34 or 37
// MasterCard uses 16-digit numbers | MasterCard numbers start with 51, 52, 53, 54, or 55
// Visa uses 13- and 16-digit | 4

File: pset6/credit/credit.py
from cs50 import get_int
from sys import exit

credit = get_int("Number: ")
length = len(str(credit))
digits = [None] * length

if length in [13, 15, 16]:
    i = length - 1
    while i >= 0:
        digits[i] = int(str(credit % 10)[:1])
        credit = credit / 10
        i -= 1

    total = 0
    i = length - 2
    while i >= 0:
        product = digits[i] * 2
        length_of_product = len(str(product))
        if length_of_product > 1:
            j = length_of_product - 1
            while j >= 0:
                total += int(str(product % 10)[:1])
                product = product / 10
                j -= 1
        else:
            total += product
        i -= 2

    i = length - 1
    while i >= 0:
        total += digits[i]
        i -= 2
    # print(total)

    if total % 10 != 0:
        print("INVALID")
        exit()

    if length == 15 and digits[0] == 3 and (digits[1] == 4 or digits[1] == 7):
        print("AMEX")
        exit()
    elif (length == 13 or length == 16) and digits[0] == 4:
        print("VISA")
        exit()
    elif length == 16 and digits[0] == 5 and (digits[1] == 1 or digits[1] == 2 or digits[1] == 3 or digits[1] == 4 or digits[1] == 5):
        print("MASTERCARD")
        exit()
    else:
        print("INVALID")
        exit()
else:
    print("INVALID")
    exit()
